src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -928,7 +989,7 @@ ByteBuffer overlapDetection(ByteBuffer src, ByteBuffer dst) {
          * allocated because for code simplicity.
          */
         byte[] overlapDetection(byte[] in, int inOfs, byte[] out, int outOfs) {
-            if (in == out && inOfs < outOfs) {
+            if (in == out && (!encryption || inOfs < outOfs)) {
```

Comment:
So, we will always allocate an output buffer for decryption if in==out? Why just decryption? Update the javadoc for this method with the reason?

Reply:
If the crypto is decryption in-place, an internal output buffer is needed in case the auth tag fails, otherwise the input buffer would be zero'ed.

Reply:
If decryption fails with a bad auth tag, the in is not overwritten because it's in-place.  Encryption is not needed because there is nothing to check.  I can add a comment.

Reply:
Hmm ok, so if it's not decryption in-place, then output buffer would still be zero'ed when the auth tag failed, but this is ok?

Reply:
This is able in-place, not about two separate buffers.. zeroing happens somewhere else for all decryption bad buffers

Reply:
Yes, I know. Basically, we are trying to optimize performance by trying to write into the supplied buffers (out) as much as we can. But then when tag verification failed, the "written" bytes are erased w/ 0. Ideal case would be not to touch the output buffer until after the tag verification succeeds. Isn't this the previous approach? Verify the tag first and then write out the plain text afterwards.

Reply:
With this new intrinsic doing both ghash and gctr at the same time, I cannot do the that ghash check first before the gctr op.  I wish I could

Reply:
Oh-well, ok.